sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,531 @@
+            // type without min/max and advanced statistics like histogram.
+            logDebug("[CBO] No range comparison statistics for String/Binary type " + attrRef)
+            None
+        }
+    }
+  }
+
+  /**
+   * This method converts a Literal value of numeric type to a BigDecimal value.
+   * In order to avoid type casting error such as Java int to Java long, we need to
+   * convert a numeric integer value to String, and then convert it to long,
+   * and then convert it to BigDecimal.
+   *
+   * @param attrDataType the column data type
+   * @param litValue the literal value
+   * @return a BigDecimal value
+   */
+  def numericLiteralToBigDecimal(
+       attrDataType: DataType,
+       litValue: Any)
+    : BigDecimal = {
+    attrDataType match {
+      case _: IntegralType =>
+        BigDecimal(litValue.toString)
+
+      case _: FractionalType =>
+        BigDecimal(litValue.toString.toDouble)
+
+      case DateType =>
+        val dateLiteral =
+          DateTimeUtils.fromJavaDate(litValue.asInstanceOf[Date]).toString
+        BigDecimal(java.lang.Long.valueOf(dateLiteral))
+
+      case TimestampType =>
+        val tsLiteral =
+          DateTimeUtils.fromJavaTimestamp(litValue.asInstanceOf[Timestamp]).toString
+        BigDecimal(java.lang.Long.valueOf(tsLiteral))
+    }
+  }
+
+  /**
+   * Returns a percentage of rows meeting an equality (=) expression.
+   * This method evaluates the equality predicate for all data types.
+   *
+   * @param attrRef an AttributeReference (or a column)
+   * @param literal a literal value (or constant)
+   * @param update a boolean flag to specify if we need to update ColumnStat of a given column
+   *               for subsequent conditions
+   * @return an optional double value to show the percentage of rows meeting a given condition
+   */
+  def evaluateEqualTo(
+      attrRef: AttributeReference,
+      literal: Literal,
+      update: Boolean)
+    : Option[Double] = {
+
+    val aColStat = mutableColStats(attrRef.exprId)
+    val ndv = aColStat.distinctCount
+
+    // decide if the value is in [min, max] of the column.
+    // We currently don't store min/max for binary/string type.
+    // Hence, we assume it is in boundary for binary/string type.
+    val statsRange = Range(aColStat.min, aColStat.max, attrRef.dataType)
+    val litRange = Range(Some(literal.value), Some(literal.value), literal.dataType)
+    val inBoundary: Boolean = Range.isIntersected(statsRange, litRange)
+
+    if (inBoundary) {
+
+      if (update) {
+        // We update ColumnStat structure after apply this equality predicate.
+        // Set distinctCount to 1.  Set nullCount to 0.
+        val newStats = attrRef.dataType match {
+          case _: NumericType =>
+            val newValue = Some(literal.value)
+            aColStat.copy(distinctCount = 1, min = newValue,
+              max = newValue, nullCount = 0)
+          case DateType =>
+            val dateValue = Some(literal.value)

```

Comment:
this is wrong, a `DateType` literal will have an int value, because the internal type of `DateType` is `Int`. We should convert int to Date here.